### PR TITLE
Converting WeakArray to a more scalable model

### PR DESCRIPTION
**Pull request by plivesey**
Now a simple TypeAlias creates a new WeakArray<SomeProtocol>
This is useful because we're about to add a new array

### CODE REVIEW

File: ConsistencyManager/DataStructures/Array+Weak.swift [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    /**
+     This defines a filter function which returns an array of WeakHolder.
+     However, the `isIncluded` function converts the inner elements (instead of the outer elements).
+
+     This allows you to write code like:
+     ```
+     var array = [WeakStruct<UIViewController>(element: UIViewController())]
+     let viewArray: [WeakStruct<UIView>] = array.flatMap { (viewController: UIViewController?) in
```
Comment (freshnitesh) [inline ConsistencyManager/DataStructures/Array+Weak.swift]: did you mean to use `filter` here?

File: ConsistencyManager/DataStructures/WeakArray.swift [hunk 2]
```diff
@@ -9,24 +9,37 @@
...
+ This class defines an array which doesn't hold strong references to its elements.
+ If an element in the array gets dealloced at some point, accessing that element will just return nil.
+ It takes advantage of out Array+Weak extension, so all the functions here are just pass throughs to the Array class.
 
  You cannot put structs into the WeakArray because structs cannot be declared with weak. You can only put classes into the array.
 
  KNOWN BUGS
 
  You can't create a WeakArray<SomeProtocol>. This is because of an Apple bug: https://bugs.swift.org/browse/SR-1176.
+ So, instead of this, it's recommended to create your own wrapper class and
```
Comment (freshnitesh) [inline ConsistencyManager/DataStructures/WeakArray.swift]: I think you forgot to finish the sentence

**Review by freshnitesh — APPROVED** [review]

File: ConsistencyManager/DataStructures/Array+Weak.swift [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        return nonOptionalElements
+    }
+
+    /**
+     This defines a map function which returns an array of WeakHolder.
+     However, the `transform` function converts the inner elements (instead of the outer elements).
+     This allows you to write code like:
+     ```
+     var array = [WeakStruct<UIViewController>(element: UIViewController())]
+        let viewArray: [WeakStruct<UIView>] = array.map { (viewController: UIViewController?) in
```
Comment (nicksnyder) [inline ConsistencyManager/DataStructures/Array+Weak.swift]: fix indentation

File: ConsistencyManager/DataStructures/Array+Weak.swift [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    /**
+     This defines a flatMap function which returns an array of WeakHolder.
+     However, the `transform` function converts the inner elements (instead of the outer elements).
+
+     IMPORTANT: This function automatically runs a `prune()` as it goes as you may expect with `flatMap`.
+     It will call the transform function on any inner element (including nil), but will filter out all nils in the final result.
+
+     This allows you to write code like:
+     ```
+     var array = [WeakStruct<UIViewController>(element: UIViewController())]
```
Comment (nicksnyder) [inline ConsistencyManager/DataStructures/Array+Weak.swift]: need an extra () at the end?

File: ConsistencyManager/DataStructures/Array+Weak.swift [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        }
+        return nonOptionalElements
+    }
+
+    /**
+     This defines a map function which returns an array of WeakHolder.
+     However, the `transform` function converts the inner elements (instead of the outer elements).
+     This allows you to write code like:
+     ```
+     var array = [WeakStruct<UIViewController>(element: UIViewController())]
```
Comment (nicksnyder) [inline ConsistencyManager/DataStructures/Array+Weak.swift]: need an extra () at the end?

File: ConsistencyManager/DataStructures/Array+Weak.swift [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        }
+    }
+
+    /**
+     This defines a filter function which returns an array of WeakHolder.
+     However, the `isIncluded` function converts the inner elements (instead of the outer elements).
+
+     This allows you to write code like:
+     ```
+     var array = [WeakStruct<UIViewController>(element: UIViewController())]
```
Comment (nicksnyder) [inline ConsistencyManager/DataStructures/Array+Weak.swift]: () at the end

File: ConsistencyManager/DataStructures/WeakArray.swift [hunk 7]
```diff
@@ -41,21 +54,25 @@ public struct WeakArray<T: AnyObject>: ExpressibleByArrayLiteral {
...
-        for element in elements {
-            data.append(weakClosureWithValue(element))
+    public init(arrayLiteral elements: T.Element?...) {
+        data = elements.map { element in
+            return T.init(element: element)
         }
     }
 
+    /**
+     Creates an array with the inner type `[WeakBox<T>]`.
```
Comment (nicksnyder) [inline ConsistencyManager/DataStructures/WeakArray.swift]: `[WeakHolder<T>]`?

**Review by nicksnyder — APPROVED** [review]

**Comment by plivesey** [comment]
Fixed all suggestions from @nicksnyder

**Pull request merged**